// @nestjs/microservices/client/client-rmq.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_service_1 = require("@nestjs/common/services/logger.service");
const load_package_util_1 = require("@nestjs/common/utils/load-package.util");
const random_string_generator_util_1 = require("@nestjs/common/utils/random-string-generator.util");
const events_1 = require("events");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const constants_1 = require("./../constants");
const client_proxy_1 = require("./client-proxy");
const constants_2 = require("./constants");
let rqmPackage = {};
class ClientRMQ extends client_proxy_1.ClientProxy {
    constructor(options) {
        super();
        this.options = options;
        this.logger = new logger_service_1.Logger(client_proxy_1.ClientProxy.name);
        this.client = null;
        this.channel = null;
        this.urls = this.getOptionsProp(this.options, 'urls') || [
            constants_1.RQM_DEFAULT_URL,
        ];
        this.queue =
            this.getOptionsProp(this.options, 'queue') ||
                constants_1.RQM_DEFAULT_QUEUE;
        this.prefetchCount =
            this.getOptionsProp(this.options, 'prefetchCount') ||
                constants_1.RQM_DEFAULT_PREFETCH_COUNT;
        this.isGlobalPrefetchCount =
            this.getOptionsProp(this.options, 'isGlobalPrefetchCount') ||
                constants_1.RQM_DEFAULT_IS_GLOBAL_PREFETCH_COUNT;
        this.queueOptions =
            this.getOptionsProp(this.options, 'queueOptions') ||
                constants_1.RQM_DEFAULT_QUEUE_OPTIONS;
        load_package_util_1.loadPackage('amqplib', ClientRMQ.name);
        rqmPackage = load_package_util_1.loadPackage('amqp-connection-manager', ClientRMQ.name);
    }
    close() {
        this.channel && this.channel.close();
        this.client && this.client.close();
    }
    consumeChannel() {
        this.channel.addSetup(channel => channel.consume(constants_2.RABBITMQ_REPLY_QUEUE, msg => this.responseEmitter.emit(msg.properties.correlationId, msg), { noAck: true }));
    }
    connect() {
        if (this.client) {
            return this.connection;
        }
        this.client = this.createClient();
        this.handleError(this.client);
        const connect$ = this.connect$(this.client);
        this.connection = this.mergeDisconnectEvent(this.client, connect$)
            .pipe(operators_1.switchMap(() => this.createChannel()), operators_1.share())
            .toPromise();
        return this.connection;
    }
    createChannel() {
        return new Promise(resolve => {
            this.channel = this.client.createChannel({
                json: false,
                setup: channel => this.setupChannel(channel, resolve),
            });
        });
    }
    createClient() {
        return rqmPackage.connect(this.urls);
    }
    mergeDisconnectEvent(instance, source$) {
        const close$ = rxjs_1.fromEvent(instance, constants_1.DISCONNECT_EVENT).pipe(operators_1.map(err => {
            throw err;
        }));
        return rxjs_1.merge(source$, close$).pipe(operators_1.first());
    }
    async setupChannel(channel, resolve) {
        await channel.assertQueue(this.queue, this.queueOptions);
        await channel.prefetch(this.prefetchCount, this.isGlobalPrefetchCount);
        this.responseEmitter = new events_1.EventEmitter();
        this.responseEmitter.setMaxListeners(0);
        this.consumeChannel();
        resolve();
    }
    publish(message, callback) {
        try {
            const correlationId = random_string_generator_util_1.randomStringGenerator();
            const listener = ({ content }) => this.handleMessage(JSON.parse(content.toString()), callback);
            this.responseEmitter.on(correlationId, listener);
            this.channel.sendToQueue(this.queue, Buffer.from(JSON.stringify(message)), {
                replyTo: constants_2.RABBITMQ_REPLY_QUEUE,
                correlationId,
            });
            return () => this.responseEmitter.removeListener(correlationId, listener);
        }
        catch (err) {
            callback({ err });
        }
    }
    handleMessage(packet, callback) {
        const { err, response, isDisposed } = packet;
        if (isDisposed || err) {
            callback({
                err,
                response: null,
                isDisposed: true,
            });
        }
        callback({
            err,
            response,
        });
    }
    handleError(client) {
        client.addListener(constants_1.ERROR_EVENT, err => this.logger.error(err));
    }
}
exports.ClientRMQ = ClientRMQ;
