// probot/lib/bin/read-env-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readEnvOptions = void 0;
const get_private_key_1 = require("@probot/get-private-key");
function readEnvOptions(env = process.env) {
    const privateKey = get_private_key_1.getPrivateKey({ env });
    return {
        args: [],
        privateKey: (privateKey && privateKey.toString()) || undefined,
        appId: Number(env.APP_ID),
        port: Number(env.PORT) || 3000,
        host: env.HOST,
        secret: env.WEBHOOK_SECRET,
        webhookPath: env.WEBHOOK_PATH,
        webhookProxy: env.WEBHOOK_PROXY_URL,
        logLevel: env.LOG_LEVEL,
        logFormat: env.LOG_FORMAT,
        logLevelInString: env.LOG_LEVEL_IN_STRING === "true",
        logMessageKey: env.LOG_MESSAGE_KEY,
        sentryDsn: env.SENTRY_DSN,
        redisConfig: env.REDIS_URL,
        baseUrl: env.GHE_HOST
            ? `${env.GHE_PROTOCOL || "https"}://${env.GHE_HOST}/api/v3`
            : "https://api.github.com",
    };
}
exports.readEnvOptions = readEnvOptions;
