// @gitbeaker/node/dist/index.es.js
import * as Gitbeaker from '@gitbeaker/core';
import { createInstance, defaultRequest as defaultRequest$1, wait, modifyServices } from '@gitbeaker/requester-utils';
import Got from 'got';
import { decamelizeKeys } from 'xcase';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

function defaultRequest(service, _a) {
    var _b = _a === void 0 ? {} : _a, body = _b.body, query = _b.query, sudo = _b.sudo, method = _b.method;
    var options = defaultRequest$1(service, { body: body, query: query, sudo: sudo, method: method });
    // FIXME: Not the best comparison, but...it will have to do for now.
    if (typeof body === 'object' && body.constructor.name !== 'FormData') {
        options.json = decamelizeKeys(body);
        delete options.body;
    }
    if (service.url.includes('https') &&
        service.rejectUnauthorized != null &&
        service.rejectUnauthorized === false) {
        options.https = {
            rejectUnauthorized: service.rejectUnauthorized,
        };
    }
    return options;
}
function processBody(_a) {
    var rawBody = _a.rawBody, headers = _a.headers;
    var contentType = headers['content-type'] || '';
    switch (contentType) {
        case 'application/json': {
            return rawBody.length === 0 ? {} : JSON.parse(rawBody.toString());
        }
        case 'application/octet-stream':
        case 'binary/octet-stream':
        case 'application/gzip': {
            return Buffer.from(rawBody);
        }
        default: {
            return rawBody.toString();
        }
    }
}
function handler(endpoint, options) {
    return __awaiter(this, void 0, void 0, function () {
        var retryCodes, maxRetries, response, i, waitTime, e_1, output, statusCode, headers, body;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    retryCodes = [429, 502];
                    maxRetries = 10;
                    i = 0;
                    _a.label = 1;
                case 1:
                    if (!(i < maxRetries)) return [3 /*break*/, 9];
                    waitTime = Math.pow(2, i) * 0.1;
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, 4, , 8]);
                    if (options.method === 'stream')
                        return [2 /*return*/, Got(endpoint, options)];
                    return [4 /*yield*/, Got(endpoint, options)];
                case 3:
                    response = _a.sent(); // eslint-disable-line
                    return [3 /*break*/, 9];
                case 4:
                    e_1 = _a.sent();
                    if (!e_1.response) return [3 /*break*/, 7];
                    if (!retryCodes.includes(e_1.response.statusCode)) return [3 /*break*/, 6];
                    return [4 /*yield*/, wait(waitTime)];
                case 5:
                    _a.sent(); // eslint-disable-line
                    return [3 /*break*/, 8]; // eslint-disable-line
                case 6:
                    if (typeof e_1.response.body === 'string' && e_1.response.body.length > 0) {
                        try {
                            output = JSON.parse(e_1.response.body);
                            e_1.description = output.error || output.message;
                        }
                        catch (err) {
                            e_1.description = e_1.response.body;
                        }
                    }
                    _a.label = 7;
                case 7: throw e_1;
                case 8:
                    i += 1;
                    return [3 /*break*/, 1];
                case 9:
                    statusCode = response.statusCode, headers = response.headers;
                    body = processBody(response);
                    return [2 /*return*/, { body: body, headers: headers, status: statusCode }];
            }
        });
    });
}
var Requester = createInstance(defaultRequest, handler);

var APIServices = modifyServices(Gitbeaker, { requester: Requester });
var  
// Groups
Groups = APIServices.Groups, GroupAccessRequests = APIServices.GroupAccessRequests, GroupBadges = APIServices.GroupBadges, GroupCustomAttributes = APIServices.GroupCustomAttributes, GroupIssueBoards = APIServices.GroupIssueBoards, GroupMembers = APIServices.GroupMembers, GroupMilestones = APIServices.GroupMilestones, GroupProjects = APIServices.GroupProjects, GroupRunners = APIServices.GroupRunners, GroupVariables = APIServices.GroupVariables, GroupLabels = APIServices.GroupLabels, Epics = APIServices.Epics, EpicIssues = APIServices.EpicIssues, EpicNotes = APIServices.EpicNotes, EpicDiscussions = APIServices.EpicDiscussions, 
// Users
Users = APIServices.Users, UserCustomAttributes = APIServices.UserCustomAttributes, UserEmails = APIServices.UserEmails, UserImpersonationTokens = APIServices.UserImpersonationTokens, UserKeys = APIServices.UserKeys, UserGPGKeys = APIServices.UserGPGKeys, 
// Projects
Branches = APIServices.Branches, Commits = APIServices.Commits, CommitDiscussions = APIServices.CommitDiscussions, ContainerRegistry = APIServices.ContainerRegistry, Deployments = APIServices.Deployments, DeployKeys = APIServices.DeployKeys, Environments = APIServices.Environments, FreezePeriods = APIServices.FreezePeriods, Issues = APIServices.Issues, IssuesStatistics = APIServices.IssuesStatistics, IssueNotes = APIServices.IssueNotes, IssueDiscussions = APIServices.IssueDiscussions, IssueAwardEmojis = APIServices.IssueAwardEmojis, Jobs = APIServices.Jobs, Labels = APIServices.Labels, MergeRequests = APIServices.MergeRequests, MergeRequestApprovals = APIServices.MergeRequestApprovals, MergeRequestAwardEmojis = APIServices.MergeRequestAwardEmojis, MergeRequestDiscussions = APIServices.MergeRequestDiscussions, MergeRequestNotes = APIServices.MergeRequestNotes, Packages = APIServices.Packages, Pipelines = APIServices.Pipelines, PipelineSchedules = APIServices.PipelineSchedules, PipelineScheduleVariables = APIServices.PipelineScheduleVariables, Projects = APIServices.Projects, ProjectAccessRequests = APIServices.ProjectAccessRequests, ProjectBadges = APIServices.ProjectBadges, ProjectCustomAttributes = APIServices.ProjectCustomAttributes, ProjectImportExport = APIServices.ProjectImportExport, ProjectIssueBoards = APIServices.ProjectIssueBoards, ProjectHooks = APIServices.ProjectHooks, ProjectMembers = APIServices.ProjectMembers, ProjectMilestones = APIServices.ProjectMilestones, ProjectSnippets = APIServices.ProjectSnippets, ProjectSnippetNotes = APIServices.ProjectSnippetNotes, ProjectSnippetDiscussions = APIServices.ProjectSnippetDiscussions, ProjectSnippetAwardEmojis = APIServices.ProjectSnippetAwardEmojis, ProtectedBranches = APIServices.ProtectedBranches, ProtectedTags = APIServices.ProtectedTags, ProjectVariables = APIServices.ProjectVariables, PushRules = APIServices.PushRules, Releases = APIServices.Releases, ReleaseLinks = APIServices.ReleaseLinks, Repositories = APIServices.Repositories, RepositoryFiles = APIServices.RepositoryFiles, Runners = APIServices.Runners, Services = APIServices.Services, Tags = APIServices.Tags, Todos = APIServices.Todos, Triggers = APIServices.Triggers, VulnerabilityFindings = APIServices.VulnerabilityFindings, 
// Genral
ApplicationSettings = APIServices.ApplicationSettings, BroadcastMessages = APIServices.BroadcastMessages, Events = APIServices.Events, FeatureFlags = APIServices.FeatureFlags, GeoNodes = APIServices.GeoNodes, GitignoreTemplates = APIServices.GitignoreTemplates, GitLabCIYMLTemplates = APIServices.GitLabCIYMLTemplates, Keys = APIServices.Keys, License = APIServices.License, LicenceTemplates = APIServices.LicenceTemplates, Lint = APIServices.Lint, Namespaces = APIServices.Namespaces, NotificationSettings = APIServices.NotificationSettings, Markdown = APIServices.Markdown, PagesDomains = APIServices.PagesDomains, Search = APIServices.Search, SidekiqMetrics = APIServices.SidekiqMetrics, Snippets = APIServices.Snippets, SystemHooks = APIServices.SystemHooks, Version = APIServices.Version, Wikis = APIServices.Wikis, 
// Bundles
GroupsBundle = APIServices.GroupsBundle, UsersBundle = APIServices.UsersBundle, ProjectsBundle = APIServices.ProjectsBundle, Gitlab = APIServices.Gitlab;

export { ApplicationSettings, Branches, BroadcastMessages, CommitDiscussions, Commits, ContainerRegistry, DeployKeys, Deployments, Environments, EpicDiscussions, EpicIssues, EpicNotes, Epics, Events, FeatureFlags, FreezePeriods, GeoNodes, GitLabCIYMLTemplates, GitignoreTemplates, Gitlab, GroupAccessRequests, GroupBadges, GroupCustomAttributes, GroupIssueBoards, GroupLabels, GroupMembers, GroupMilestones, GroupProjects, GroupRunners, GroupVariables, Groups, GroupsBundle, IssueAwardEmojis, IssueDiscussions, IssueNotes, Issues, IssuesStatistics, Jobs, Keys, Labels, LicenceTemplates, License, Lint, Markdown, MergeRequestApprovals, MergeRequestAwardEmojis, MergeRequestDiscussions, MergeRequestNotes, MergeRequests, Namespaces, NotificationSettings, Packages, PagesDomains, PipelineScheduleVariables, PipelineSchedules, Pipelines, ProjectAccessRequests, ProjectBadges, ProjectCustomAttributes, ProjectHooks, ProjectImportExport, ProjectIssueBoards, ProjectMembers, ProjectMilestones, ProjectSnippetAwardEmojis, ProjectSnippetDiscussions, ProjectSnippetNotes, ProjectSnippets, ProjectVariables, Projects, ProjectsBundle, ProtectedBranches, ProtectedTags, PushRules, ReleaseLinks, Releases, Repositories, RepositoryFiles, Runners, Search, Services, SidekiqMetrics, Snippets, SystemHooks, Tags, Todos, Triggers, UserCustomAttributes, UserEmails, UserGPGKeys, UserImpersonationTokens, UserKeys, Users, UsersBundle, Version, VulnerabilityFindings, Wikis };
