// @nestjs/microservices/server/server-rmq.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const load_package_util_1 = require("@nestjs/common/utils/load-package.util");
const constants_1 = require("../constants");
const server_1 = require("./server");
let rqmPackage = {};
class ServerRMQ extends server_1.Server {
    constructor(options) {
        super();
        this.options = options;
        this.server = null;
        this.channel = null;
        this.urls = this.getOptionsProp(this.options, 'urls') || [
            constants_1.RQM_DEFAULT_URL,
        ];
        this.queue =
            this.getOptionsProp(this.options, 'queue') ||
                constants_1.RQM_DEFAULT_QUEUE;
        this.prefetchCount =
            this.getOptionsProp(this.options, 'prefetchCount') ||
                constants_1.RQM_DEFAULT_PREFETCH_COUNT;
        this.isGlobalPrefetchCount =
            this.getOptionsProp(this.options, 'isGlobalPrefetchCount') ||
                constants_1.RQM_DEFAULT_IS_GLOBAL_PREFETCH_COUNT;
        this.queueOptions =
            this.getOptionsProp(this.options, 'queueOptions') ||
                constants_1.RQM_DEFAULT_QUEUE_OPTIONS;
        load_package_util_1.loadPackage('amqplib', ServerRMQ.name);
        rqmPackage = load_package_util_1.loadPackage('amqp-connection-manager', ServerRMQ.name);
    }
    async listen(callback) {
        await this.start(callback);
    }
    close() {
        this.channel && this.channel.close();
        this.server && this.server.close();
    }
    async start(callback) {
        this.server = this.createClient();
        this.server.on(constants_1.CONNECT_EVENT, _ => {
            this.channel = this.server.createChannel({
                json: false,
                setup: channel => this.setupChannel(channel, callback),
            });
        });
        this.server.on(constants_1.DISCONNECT_EVENT, err => {
            this.logger.error(constants_1.DISCONNECTED_RMQ_MESSAGE);
        });
    }
    createClient() {
        return rqmPackage.connect(this.urls);
    }
    async setupChannel(channel, callback) {
        await channel.assertQueue(this.queue, this.queueOptions);
        await channel.prefetch(this.prefetchCount, this.isGlobalPrefetchCount);
        channel.consume(this.queue, msg => this.handleMessage(msg), {
            noAck: true,
        });
        callback();
    }
    async handleMessage(message) {
        const { content, properties } = message;
        const packet = JSON.parse(content.toString());
        const pattern = JSON.stringify(packet.pattern);
        const handler = this.getHandlerByPattern(pattern);
        if (!handler) {
            const status = 'error';
            return this.sendMessage({ status, err: constants_1.NO_PATTERN_MESSAGE }, properties.replyTo, properties.correlationId);
        }
        const response$ = this.transformToObservable(await handler(packet.data));
        const publish = data => this.sendMessage(data, properties.replyTo, properties.correlationId);
        response$ && this.send(response$, publish);
    }
    sendMessage(message, replyTo, correlationId) {
        const buffer = Buffer.from(JSON.stringify(message));
        this.channel.sendToQueue(replyTo, buffer, { correlationId });
    }
}
exports.ServerRMQ = ServerRMQ;
